### PR TITLE
fix(outputs.iotdb): Handle paths that contain illegal characters

### LINKED ISSUE #14518: IoTDB device_id [ILLEGAL_PATH(509)] Exception occurred: insertRecord failed
### Relevant telegraf.conf

```toml
[[inputs.win_perf_counters]]
  [[inputs.win_perf_counters.object]]
    ObjectName = "Network Interface"
    Counters = ["Bytes Received/sec","Bytes Sent/sec","Packets Received/sec","Packets Sent/sec"]
    Instances = ["*"]
    Measurement = "root.WinPerfCounters.$COMPUTERNAME.WinNet"

[[outputs.iotdb]]
  # host, port, user, password have been purposefully omitted
  # to avoid exposing sensitive data
  host = ""
  port = ""
  user = ""
  password = ""
  uint64_conversion = "int64_clip"
  timestamp_precision = "millisecond"
  convert_tags_to = "device_id"
  timeout = "0s"
```

### Logs from Telegraf

```text
2023/12/05 14:59:21 [ERROR]  Error Code: 509, Message: [ILLEGAL_PATH(509)] Exception occurred: insertRecord failed. root.WinPerfCounters.DC5.WinNet.Adattatore Ethernet vmxnet3 is not a legal path
2023/12/05 14:59:22 [ERROR]  Error Code: 509, Message: [ILLEGAL_PATH(509)] Exception occurred: insertRecord failed. root.WinPerfCounters.DC5.WinNet.isatap.{DAACFE3D-0417-4B20-8F47-9599B4C94381} is not a legal path
```

### System info

Telegraf 1.29.1, Windows Server 2019

### Docker

_No response_

### Steps to reproduce

1. Configure telegraf to use IoTDB output
2. Set convert_tag_to = "device_id"
3. Run telegraf in debug mode

### Expected behavior

The path should be corrected by enclosing the tags wich contain a character not supported by IoTDB in ``

### Actual behavior

Telegraf simply tryes to add the path as it is and dosen't give any way to correct such behaviour

### Additional info

Host, port, user, password have been purposefully omitted to avoid exposing sensitive data.

We got a little working fix which is running on our production sites, and are willing to share the current code to get some reviews and maybe improvements.

### PR DESCRIPTION
**Pull request by giovanni-bellini-argo**
## Summary

Support for IoTDB ILLEGAL_PATH error when tags contains an illegal character, explained more in detail in issue.

## Checklist

- [x] No AI generated code was used in this PR

## Related issues

resolves #14518

### CODE REVIEW

**Review by powersj — COMMENTED** [review]
Thanks for the PR!

Looking at the [IoTBD](https://iotdb.apache.org/UserGuide/V0.13.x/Reference/Syntax-Conventions.html#identifiers) this does appear to be the suggested way of handling special characters in the path.

Can you break out the regex checking into a new function and please add tests for that function with both cases including special characters? My only concern with these types of changes is future changes down the road or corner cases :)

Thanks!

**Comment by srebhan** [comment]
@giovanni-bellini-argo please also sign the CLA!

**Comment by giovanni-bellini-argo** [comment]
!signed-cla

**Review by powersj — COMMENTED** [review]
Can you confirm my understanding after playing with this a bit more: A path is made up of three elements:

* root - reserved word
* storage group
* child - one or more items separated by periods

The storage group can only have:

> The storage group name can only be characters, numbers and underscores.

I think this is what you are currently testing for. However, the child elements seem to allow a greater set of elements. This means the following are valid without escaping them.

> root.foo.3
> root.foo.()

Is that your understanding as well? I've been trying this with version 0.13.0, so that may play into this as well.

**Comment by giovanni-bellini-argo** [comment]
We are testing on V1.2.2 and could not create child nodes with any other characters outside of underscores and alphanumeric characters

**Comment by powersj** [comment]
Thanks!

> We are testing on V1.2.2 and could not create child nodes with any other characters outside of underscores and alphanumeric characters.

I am glad we are trying on different versions. Unfortunately, having different behavior on different versions does mean we would need to make this opt-in as well. We cannot be breaking existing users on versions that do not run into this.

How we typically handle this is add a config option, something like:

```toml
# Mode to sanitize tags
# By default, tags are not sanitized. Possible options:
#   1.2 - only allows alpha/numeric and underscores, and non-numeric values. Otherwise, 
#         backticks are applied
sanitize_tag = ""
```

Then if that value is set to "1.2", you can apply your sanitization function.

I would still like to see a specific test as well that only tests the sanitize function with a various test cases.

Does that make sense?

**Comment by powersj** [comment]
I've also asked over in the iotdb repo for guidance on valid characters:

https://github.com/apache/iotdb/discussions/11840

**Comment by giovanni-bellini-argo** [comment]
a list of updates:
1. added the option as per your request and made some researches: from version 1.x.x they completely changed the handling of nodes names which can apparently take only letters, numbers and underscores but can't be composed of only numbers.
2. made some more checks to throw an error if root (which is a reserved kayword) is used and if ` is misused
3. added the specific tests cases.

**Comment by powersj** [comment]
Thanks for the updates and doing that research. Almost there!

1. I do not see your name in our CLA signed sheet. Can you please sign it?
2. Looks like there a couple in lint issues:

```s
plugins/outputs/iotdb/iotdb.go:277:28  gosimple  S1007: should use raw string (`...`) with regexp.Compile to avoid having to escape twice
plugins/outputs/iotdb/iotdb.go:310:5   revive    var-naming: don't use underscores in Go names; var tag_value should be tagValue
```

3. And a test issue with your new tests, specifically: `TestTagsHandling/treat_tags_as_device_IDs`

4. Please update the sample config and README with the new config option for `sanitize_tag`.

Happy to help with any of these updates, let me know if you do want assistence.

**Comment by giovanni-bellini-argo** [comment]
as per the CLA i did sign it on google (first time doing this so i cheched out online and found most use the google db) i did also read your guide about contributions and followed the link to sign the CLA on your site but the form has some iessues:

image

**Comment by powersj** [comment]
Hey @giovanni-bellini-argo,

That seems to be your browser saying it is blocked by your content security policy. Can you please try a different browser?

Thanks!

File: plugins/outputs/iotdb/README.md [hunk 1]
```diff
@@ -141,4 +141,19 @@ to use them.
   ##   - "fields"     --  root.sg.device, s1=100, s2="hello", tag1="private", tag2="working"
   ##   - "device_id"  --  root.sg.device.private.working, s1=100, s2="hello"
   # convert_tags_to = "device_id"
+
+  ## Handling of Unsupported Characters
+  ## Some characters in different versions of IoTDB are not supported in path name
+  ## A guide with suggetions on valid paths can be found here:
+  ## for iotdb 0.13.x           -> https://iotdb.apache.org/UserGuide/V0.13.x/Reference/Syntax-Conventions.html#timestamp-literals
```
Comment (powersj) [inline plugins/outputs/iotdb/README.md:148]: Should this link point to the identifiers instead:

https://iotdb.apache.org/UserGuide/V0.13.x/Reference/Syntax-Conventions.html#identifiers

File: plugins/outputs/iotdb/iotdb.go [hunk 2]
```diff
@@ -229,6 +231,60 @@ func (s *IoTDB) convertMetricsToRecordsWithTags(metrics []telegraf.Metric) (*rec
...
+func (s *IoTDB) validateTag(tag string) (string, error) {
+	// matches any word that has a non valid backtick
+	// `word`  							 <- dosen't match
+	// ``word , `wo`rd` , `word , word`  <- match
+	forbiddenBacktick, _ := regexp.Compile("^[^\x60].*?[\x60]+.*?[^\x60]$|^[\x60].*[\x60]+.*[\x60]$|^[\x60]+.*[^\x60]$|^[^\x60].*[\x60]+$")
+	allowedBacktick, _ := regexp.Compile("^[\x60].*[\x60]$")
+
+	// IoTDB uses "root" as a keyword and can be called only at the start of the path
+	if tag == "root" {
+		return "", fmt.Errorf("can't append keyword \"root\" to IoTDB base path")
```
Comment (powersj) [inline plugins/outputs/iotdb/iotdb.go:258]: Should this error message actually read that you "cannot use \"root\" as tag?

```suggestion
		return "", errors.New("cannot append keyword \"root\" to IoTDB base path")
```

File: plugins/outputs/iotdb/iotdb.go [hunk 3]
```diff
@@ -229,6 +231,60 @@ func (s *IoTDB) convertMetricsToRecordsWithTags(metrics []telegraf.Metric) (*rec
...
+	} else if forbiddenBacktick.MatchString(tag) { // returns an error if the backsticks are used in an inappropriate way
+		return "", fmt.Errorf("can't use ` in IoTDB base path")
```
Comment (powersj) [inline plugins/outputs/iotdb/iotdb.go:260]:
```suggestion
		return "", errors.New("cannot use ` in IoTDB base path")
```

File: plugins/outputs/iotdb/iotdb.go [hunk 4]
```diff
@@ -251,7 +307,13 @@ func (s *IoTDB) modifyRecordsWithTags(rwt *recordsWithTags) error {
...
 			topic := []string{rwt.DeviceIDList[index]}
 			for _, tag := range tags { // for each tag, append it's Value
-				topic = append(topic, tag.Value)
+				tagValue, err := s.validateTag(tag.Value) // validates tag
+
+				if err != nil {
+					return err
+				}
+
+				topic = append(topic, tagValue)
```
Comment (powersj) [inline plugins/outputs/iotdb/iotdb.go:316]:
```suggestion
				tagValue, err := s.validateTag(tag.Value) // validates tag
				if err != nil {
					return err
				}
				topic = append(topic, tagValue)
```

File: plugins/outputs/iotdb/sample.conf [hunk 5]
```diff
@@ -43,3 +43,18 @@
   ##   - "fields"     --  root.sg.device, s1=100, s2="hello", tag1="private", tag2="working"
   ##   - "device_id"  --  root.sg.device.private.working, s1=100, s2="hello"
   # convert_tags_to = "device_id"
+
+  ## Handling of Unsupported Characters
+  ## Some characters in different versions of IoTDB are not supported in path name
+  ## A guide with suggetions on valid paths can be found here:
+  ## for iotdb 0.13.x           -> https://iotdb.apache.org/UserGuide/V0.13.x/Reference/Syntax-Conventions.html#timestamp-literals
```
Comment (powersj) [inline plugins/outputs/iotdb/sample.conf:50]: Same question here about the URL anchor.

File: plugins/outputs/iotdb/iotdb_test.go [hunk 6]
```diff
@@ -271,6 +271,68 @@ func TestMetricConversionToRecordsWithTags(t *testing.T) {
...
+			input:    []string{"wo rd", "@", "$", "#", ":", "{", "}", "1", "1234"},
+		},
+		{ //test on forbidden word and forbidden syntax
+			name:     "Errors",
+			plugin:   func() *IoTDB { s := newIoTDB(); s.SanitizeTags = "1.3"; return s }(),
+			expected: []string{"", ""},
+			input:    []string{"root", "wo`rd"},
+		},
+
+		{
```
Comment (powersj) [inline plugins/outputs/iotdb/iotdb_test.go:301]:
```suggestion
		},
		{
```

**Review by powersj — COMMENTED** [review]
Sorry for the delay, thanks again for the re-work, I think we are nearly there!

**Review by powersj — APPROVED** [review]
Thanks for working through this!

File: plugins/outputs/iotdb/README.md [hunk 7]
```diff
@@ -141,4 +141,19 @@ to use them.
   ##   - "fields"     --  root.sg.device, s1=100, s2="hello", tag1="private", tag2="working"
   ##   - "device_id"  --  root.sg.device.private.working, s1=100, s2="hello"
   # convert_tags_to = "device_id"
+
+  ## Handling of Unsupported Characters
```
Comment (srebhan) [inline plugins/outputs/iotdb/README.md:145]:
```suggestion
  ## Handling of unsupported characters
```

File: plugins/outputs/iotdb/iotdb.go [hunk 8]
```diff
@@ -229,6 +231,60 @@ func (s *IoTDB) convertMetricsToRecordsWithTags(metrics []telegraf.Metric) (*rec
...
+	return tag
+}
+
+// checks is the tag contains any IoTDB invalid character
+func (s *IoTDB) validateTag(tag string) (string, error) {
+	// matches any word that has a non valid backtick
+	// `word`  							 <- dosen't match
+	// ``word , `wo`rd` , `word , word`  <- match
+	forbiddenBacktick, _ := regexp.Compile("^[^\x60].*?[\x60]+.*?[^\x60]$|^[\x60].*[\x60]+.*[\x60]$|^[\x60]+.*[^\x60]$|^[^\x60].*[\x60]+$")
+	allowedBacktick, _ := regexp.Compile("^[\x60].*[\x60]$")
```
Comment (srebhan) [inline plugins/outputs/iotdb/iotdb.go:254]: If we know this should compile, you should do
```suggestion
	forbiddenBacktick := regexp.MustCompile("^[^\x60].*?[\x60]+.*?[^\x60]$|^[\x60].*[\x60]+.*[\x60]$|^[\x60]+.*[^\x60]$|^[^\x60].*[\x60]+$")
	allowedBacktick := regexp.MustCompile("^[\x60].*[\x60]$")
```
to catch programming errors. We might also want define those two as package-global variables to not compile them for every metric again as they seem constant...

File: plugins/outputs/iotdb/iotdb.go [hunk 9]
```diff
@@ -229,6 +231,60 @@ func (s *IoTDB) convertMetricsToRecordsWithTags(metrics []telegraf.Metric) (*rec
...
+
+	// IoTDB uses "root" as a keyword and can be called only at the start of the path
+	if tag == "root" {
+		return "", errors.New("cannot use \"root\" as tag")
```
Comment (srebhan) [inline plugins/outputs/iotdb/iotdb.go:258]:
```suggestion
		return "", errors.New("cannot use 'root' as tag")
```

File: plugins/outputs/iotdb/iotdb.go [hunk 10]
```diff
@@ -229,6 +231,60 @@ func (s *IoTDB) convertMetricsToRecordsWithTags(metrics []telegraf.Metric) (*rec
...
+	case "0.13":
+		matchUnsopportedCharacter, _ := regexp.Compile("[^0-9a-zA-Z_:@#${}\x60]")
+
+		regex := []*regexp.Regexp{matchUnsopportedCharacter}
+		regexArray = append(regexArray, regex...)
+
+	// from version 1.x.x IoTDB changed the allowed keys in nodes
+	case "1.0", "1.1", "1.2", "1.3":
+		matchUnsopportedCharacter, _ := regexp.Compile("[^0-9a-zA-Z_\x60]")
+		matchNumericString, _ := regexp.Compile(`^\d+$`)
```
Comment (srebhan) [inline plugins/outputs/iotdb/iotdb.go:277]: Same here, use `MustCompile` and initialize `regexArray` in `Init()`!

File: plugins/outputs/iotdb/sample.conf [hunk 11]
```diff
@@ -43,3 +43,18 @@
   ##   - "fields"     --  root.sg.device, s1=100, s2="hello", tag1="private", tag2="working"
   ##   - "device_id"  --  root.sg.device.private.working, s1=100, s2="hello"
   # convert_tags_to = "device_id"
+
+  ## Handling of Unsupported Characters
```
Comment (srebhan) [inline plugins/outputs/iotdb/sample.conf:47]:
```suggestion
  ## Handling of unsupported characters
```

**Review by srebhan — COMMENTED** [review]
Thanks for this nice PR @giovanni-bellini-argo! I have some small comments, which moving the regexp-compilation out of the `Write()` path being the most important one...

File: plugins/outputs/iotdb/iotdb.go (hunk 10 above)
Comment (giovanni-bellini-argo) [reply plugins/outputs/iotdb/iotdb.go:277]: why initializing `regexArray` in `Init()`?
u think it would be better to include it in the struct for future changes?

and i would also delete those useless variables and directly append the regex to the array.
```golang
regexArray := []*regexp.Regexp{} // array of compiled regex patterns

	switch s.SanitizeTags {
	case "0.13":
		regex := []*regexp.Regexp{
			regexp.MustCompile("[^0-9a-zA-Z_:@#${}\x60]"),
		}
		regexArray = append(regexArray, regex...)

	// from version 1.x.x IoTDB changed the allowed keys in nodes
	case "1.0", "1.1", "1.2", "1.3":
		regex := []*regexp.Regexp{
			regexp.MustCompile("[^0-9a-zA-Z_\x60]"),
			regexp.MustCompile(`^\d+$`),
		}

		regexArray = append(regexArray, regex...)
	default:
		return tag, nil
	}
```
the idea is to make it easy to undertand and modify

File: plugins/outputs/iotdb/iotdb.go [hunk 12]
```diff
@@ -229,6 +231,60 @@ func (s *IoTDB) convertMetricsToRecordsWithTags(metrics []telegraf.Metric) (*rec
...
+	// from version 1.x.x IoTDB changed the allowed keys in nodes
+	case "1.0", "1.1", "1.2", "1.3":
+		matchUnsopportedCharacter := regexp.MustCompile("[^0-9a-zA-Z_\x60]")
+		matchNumericString := regexp.MustCompile(`^\d+$`)
+
+		regex := []*regexp.Regexp{matchUnsopportedCharacter, matchNumericString}
+		regexArray = append(regexArray, regex...)
+	default:
+		return tag, nil
+	}
```
Comment (srebhan) [inline plugins/outputs/iotdb/iotdb.go:283]: Please move this part to the `Init()` function and store `regexArray` (maybe name it `sanityRegex`) in the `IoTDB` structure. This way, we don't need to do this for every metric we write but only once!

File: plugins/outputs/iotdb/iotdb.go [hunk 13]
```diff
@@ -229,6 +231,60 @@ func (s *IoTDB) convertMetricsToRecordsWithTags(metrics []telegraf.Metric) (*rec
...
+	return tag
+}
+
+// checks is the tag contains any IoTDB invalid character
+func (s *IoTDB) validateTag(tag string) (string, error) {
+	// matches any word that has a non valid backtick
+	// `word`  							 <- dosen't match
+	// ``word , `wo`rd` , `word , word`  <- match
+	forbiddenBacktick := regexp.MustCompile("^[^\x60].*?[\x60]+.*?[^\x60]$|^[\x60].*[\x60]+.*[\x60]$|^[\x60]+.*[^\x60]$|^[^\x60].*[\x60]+$")
+	allowedBacktick := regexp.MustCompile("^[\x60].*[\x60]$")
```
Comment (srebhan) [inline plugins/outputs/iotdb/iotdb.go:254]: Please move this outside of any function, similar to how `sampleConfig` is declared. This way, we need to compile the **static** expressions only once and not for every metric!

File: plugins/outputs/iotdb/iotdb.go [hunk 14]
```diff
@@ -229,6 +231,60 @@ func (s *IoTDB) convertMetricsToRecordsWithTags(metrics []telegraf.Metric) (*rec
...
+	// loops through all the regex patterns and if one
+	// pattern matches returns the tag between `
+	for _, regex := range regexStrings {
+		if regex.MatchString(tag) {
+			return "`" + tag + "`"
+		}
+	}
+
+	return tag
+}
```
Comment (srebhan) [inline plugins/outputs/iotdb/iotdb.go:246]: You should really merge this into `validateTag`...

**Review by srebhan — COMMENTED** [review]
Thanks for your update @giovanni-bellini-argo! Please try to compile the regular-expressions only once as suggested in the code! If you need help, I can provide a patch...

File: plugins/outputs/iotdb/iotdb.go [hunk 15]
```diff
@@ -30,9 +37,11 @@ type IoTDB struct {
 	ConvertUint64To string          `toml:"uint64_conversion"`
 	TimeStampUnit   string          `toml:"timestamp_precision"`
 	TreatTagsAs     string          `toml:"convert_tags_to"`
+	SanitizeTags    string          `toml:"sanitize_tag"`
 	Log             telegraf.Logger `toml:"-"`
 
-	session *client.Session
+	SanityRegex []*regexp.Regexp
```
Comment (srebhan) [inline plugins/outputs/iotdb/iotdb.go:43]: Please do not export this field as it is (and should be) only used internally
```suggestion
	sanityRegex []*regexp.Regexp
```

File: plugins/outputs/iotdb/iotdb.go [hunk 16]
```diff
@@ -74,6 +83,22 @@ func (s *IoTDB) Init() error {
 		s.Password = config.NewSecret([]byte("root"))
 	}
 
+	switch s.SanitizeTags {
+	case "0.13":
+		matchUnsopportedCharacter := regexp.MustCompile("[^0-9a-zA-Z_:@#${}\x60]")
+
+		regex := []*regexp.Regexp{matchUnsopportedCharacter}
+		s.SanityRegex = append(s.SanityRegex, regex...)
```
Comment (srebhan) [inline plugins/outputs/iotdb/iotdb.go:91]:
```suggestion
		s.sanityRegex = append(s.sanityRegex, matchUnsupportedCharacter)
```

File: plugins/outputs/iotdb/iotdb.go [hunk 17]
```diff
@@ -74,6 +83,22 @@ func (s *IoTDB) Init() error {
...
+
+	// from version 1.x.x IoTDB changed the allowed keys in nodes
+	case "1.0", "1.1", "1.2", "1.3":
+		matchUnsopportedCharacter := regexp.MustCompile("[^0-9a-zA-Z_\x60]")
+		matchNumericString := regexp.MustCompile(`^\d+$`)
+
+		regex := []*regexp.Regexp{matchUnsopportedCharacter, matchNumericString}
+		s.SanityRegex = append(s.SanityRegex, regex...)
```
Comment (srebhan) [inline plugins/outputs/iotdb/iotdb.go:99]:
```suggestion
		s.sanityRegex = append(s.sanityRegex, matchUnsupportedCharacter, matchNumericString)
```

File: plugins/outputs/iotdb/iotdb.go [hunk 18]
```diff
@@ -74,6 +83,22 @@ func (s *IoTDB) Init() error {
...
+	switch s.SanitizeTags {
+	case "0.13":
+		matchUnsopportedCharacter := regexp.MustCompile("[^0-9a-zA-Z_:@#${}\x60]")
+
+		regex := []*regexp.Regexp{matchUnsopportedCharacter}
+		s.SanityRegex = append(s.SanityRegex, regex...)
+
+	// from version 1.x.x IoTDB changed the allowed keys in nodes
+	case "1.0", "1.1", "1.2", "1.3":
+		matchUnsopportedCharacter := regexp.MustCompile("[^0-9a-zA-Z_\x60]")
```
Comment (srebhan) [inline plugins/outputs/iotdb/iotdb.go:95]:
```suggestion
		matchUnsupportedCharacter := regexp.MustCompile("[^0-9a-zA-Z_\x60]")
```

File: plugins/outputs/iotdb/iotdb.go [hunk 19]
```diff
@@ -74,6 +83,22 @@ func (s *IoTDB) Init() error {
 		s.Password = config.NewSecret([]byte("root"))
 	}
 
+	switch s.SanitizeTags {
+	case "0.13":
+		matchUnsopportedCharacter := regexp.MustCompile("[^0-9a-zA-Z_:@#${}\x60]")
```
Comment (srebhan) [inline plugins/outputs/iotdb/iotdb.go:88]:
```suggestion
		matchUnsupportedCharacter := regexp.MustCompile("[^0-9a-zA-Z_:@#${}\x60]")
```

File: plugins/outputs/iotdb/iotdb.go [hunk 20]
```diff
@@ -229,6 +254,28 @@ func (s *IoTDB) convertMetricsToRecordsWithTags(metrics []telegraf.Metric) (*rec
...
+		return "", errors.New("cannot use 'root' as tag")
+	} else if forbiddenBacktick.MatchString(tag) { // returns an error if the backsticks are used in an inappropriate way
+		return "", errors.New("cannot use ` in tag names")
+	} else if allowedBacktick.MatchString(tag) { // if the tag in already enclosed in tags returns the tag
+		return tag, nil
+	}
+
+	// loops through all the regex patterns and if one
+	// pattern matches returns the tag between `
+	for _, regex := range s.SanityRegex {
```
Comment (srebhan) [inline plugins/outputs/iotdb/iotdb.go:270]:
```suggestion
	for _, regex := range s.sanityRegex {
```

**Review by srebhan — COMMENTED** [review]
Looks good @giovanni-bellini-argo! I would just un-export the member variable and fix the typos and we are good to go. Tried to do it in my suggestions to ease your life but no sure I got all of them... :-)

**Comment by giovanni-bellini-argo** [comment]
done everything, thanks for the patience and help :sunglasses:

**Review by srebhan — APPROVED** [review]
Thank you for your work and patience with my requests @giovanni-bellini-argo! :-)
